### PR TITLE
chore(stdlib)!: Use default arguments in more of stdlib

### LINKED ISSUE #697: BOM arguments to `String.encode` and `String.decode` should be optional
Blocked by #388, but it's rather frustrating to always need to specify these.

Further context: #683

### PR DESCRIPTION
**Pull request by alex-snezhko**
Makes use of default arguments in various places of the stdlib where I thought they made sense.

I was also considering combining `unwrap` and `expect` in `Option`/`Result` but was unsure if I should.

Closes #697

### CODE REVIEW

File: stdlib/map.gr [hunk 1]
```diff
@@ -32,29 +32,25 @@ abstract record Map<k, v> {
...
 
 /**
- * Creates a new empty map with an initial storage of the given size. As values are added or removed, the internal storage may grow or shrink. Generally, you won't need to care about the storage size of your map and can use `Map.make()` instead.
+ * Creates a new empty map with an initial storage of the given size. As
+ * values are added or removed, the internal storage may grow or shrink.
+ * Generally, you won't need to care about the storage size of your map and
+ * can use the default size.
  *
- * @param size: The initial storage size of the map
+ * @param size: The initial storage size of the map (default is 16)
```
Comment (phated) [inline stdlib/map.gr:40]: Let's just fix the docgen so we don't need to specify this

File: stdlib/priorityqueue.gr [hunk 2]
```diff
@@ -220,12 +207,13 @@ provide let drain = pq => {
...
  * @param array: An array of values used to initialize the priority queue
- * @param comp: A comparator function used to assign priority to elements
+ * @param comp: A comparator function used to assign priority to elements (default is `compare` pervasive)
  * @returns A priority queue containing the elements from the array
  *
  * @since v0.5.4
+ * @history v0.6.0: Made `comp` a default argument
  */
-provide let fromArray = (array, comp) => {
+provide let fromArray = (array, comp=compare) => {
```
Comment (phated) [inline stdlib/priorityqueue.gr:216]: Let's rename all these "comp" to compare

**Review by phated — CHANGES_REQUESTED** [review]
Quick skim, 2 things jumped out:
1. graindoc needs to be fixed and you shouldn't be embedding defaults into the descriptions
2. No shortened names for defaulted params, we want to be very clear with their names.

File: stdlib/priorityqueue.gr (hunk 2 above)
Comment (alex-snezhko) [reply stdlib/priorityqueue.gr]: You pointing this out also brought to my attention that the sort comparator could also be optional

**Review by spotandjake — COMMENTED** [review]
I didnt look over any of the test changes just the code changes but I am liking the look of this pr.

**Comment by phated** [comment]
blocked on #1776

**Comment by alex-snezhko** [comment]
#1776 merged, no longer blocked

**Review by spotandjake — APPROVED** [review]
This looks good to me.

File: stdlib/array.gr [hunk 3]
```diff
@@ -946,16 +946,18 @@ provide let slice = (start, end=length(array), array) => {
...
  * Ordering is calculated using a comparator function which takes two array elements and must return 0 if both are equal, a positive number if the first is greater, and a negative number if the first is smaller.
- * @param comp: The comparator function used to indicate sort order
  * @param array: The array to be sorted
+ * @param compare: The comparator function used to indicate sort order
+ * 
  * @since v0.4.5
+ * @history v0.6.0: Made `compare` a default argument
  */
-provide let sort = (comp, array) => {
+provide let sort = (array, compare=compare) => {
```
Comment (ospencer) [inline stdlib/array.gr:955]: Why did you change the argument order?

Comment (alex-snezhko) [reply stdlib/array.gr:955]: No reason in particular other that I prefer default arguments appearing after required ones in the definition. I can change it back though

Comment (ospencer) [reply stdlib/array.gr]: I don't know if it's the OCamler in me or what, but I prefer the default arguments before the required ones. It's probably the OCamler in me.

Comment (phated) [reply stdlib/array.gr]: I love them at the end because it sorts them to the bottom of the documentation. So yeah, I think it's the ocamler in you 😂

Comment (phated) [reply stdlib/array.gr]: Additionally, default arguments at the end are familiar to JavaScript/TypeScript folks.

Comment (alex-snezhko) [reply stdlib/array.gr]: I think as long as we're doing data-last in the rest of the stdlib then it makes sense to have the `compare` arg before the array/list. This is kinda the same issue/argument as https://github.com/grain-lang/grain/issues/671

File: stdlib/string.gr [hunk 4]
```diff
@@ -1425,29 +1409,18 @@ let encodeHelp = (string: String, encoding: Encoding, includeBom: Bool) => {
 }
 
 /**
- * Encodes the given string using the given encoding scheme, excluding any byte-order marker.
+ * Encodes the given string using the given encoding scheme.
  *
  * @param string: The input string
  * @param encoding: The encoding to use
+ * @param includeBom: Whether or not to include a byte order marker (false by default)
```
Comment (phated) [inline stdlib/string.gr:1416]: We should add a `@default` attribute to graindoc that allows us to document defaults in plain language to avoid parentheticals like this. I'm not going to block on it though.

**Review by phated — APPROVED** [review]
This is an absolute delight! Thanks @alex-snezhko 🎉

**Comment by phated** [comment]
@alex-snezhko can you resolve the conflicts please?

File: stdlib/string.gr (hunk 4 above)
Comment (phated) [reply stdlib/string.gr:1416]: #1991
